counting/etcd.go

```diff
@@ -9,71 +9,88 @@ import (
 	"context"
 	"fmt"
 	"strconv"
-	"time"
 
 	etcdclientv3 "go.etcd.io/etcd/client/v3"
 )
 
 type Counter interface {
 	// Applies by the given delta (+ or -) and return the updated value.
-	// Count has a ttl calculated using the next tick's time.
 	// Returns (updated value, true if value was updated in memory, err if any error happened)
 	// It is possible that the value is updated but an error occurred while trying to persist it.
-	Add(context.Context, int, time.Time) (int, bool, error)
+	Increment(context.Context, int) (int, bool, error)
+	Refresh(context.Context) (int, error)
+	Delete(context.Context) error
 }
 
 // It keeps a cache of the value and updates async.
 // It works assuming there cannot be two concurrent writes to the same key.
 // Concurrency is handled at the job level, which makes this work.
 type etcdcounter struct {
-	etcdclient *etcdclientv3.Client
-	key        string
+	etcdclient   *etcdclientv3.Client
+	key          string
+	initialValue int
 
-	loaded    bool
-	value     int
-	ttlOffset time.Duration
+	loaded bool
+	value  int
 }
 
-func NewEtcdCounter(c *etcdclientv3.Client, key string, ttlOffset time.Duration) Counter {
+func NewEtcdCounter(c *etcdclientv3.Client, key string, initialValue int) Counter {
 	return &etcdcounter{
-		etcdclient: c,
-		key:        key,
-		ttlOffset:  ttlOffset,
+		etcdclient:   c,
+		initialValue: initialValue,
+		key:          key,
+		value:        initialValue,
 	}
 }
 
-func (c *etcdcounter) Add(ctx context.Context, delta int, next time.Time) (int, bool, error) {
+func (c *etcdcounter) Increment(ctx context.Context, delta int) (int, bool, error) {
 	if !c.loaded {
 		// First, load the key's value.
-		res, err := c.etcdclient.KV.Get(ctx, c.key)
+		_, err := c.Refresh(ctx)
 		if err != nil {
-			return 0, false, err
+			return c.value, false, err
 		}
-		if len(res.Kvs) == 0 {
-			c.value = 0
-			c.loaded = true
-		} else {
-			if res.Kvs[0].Value == nil {
-				return 0, false, fmt.Errorf("nil value for key %s", c.key)
-			}
-			if len(res.Kvs[0].Value) == 0 {
-				return 0, false, fmt.Errorf("empty value for key %s", c.key)
-			}
+		c.loaded = true
+	}
 
-			c.value, err = strconv.Atoi(string(res.Kvs[0].Value))
-			if err != nil {
-				return 0, false, err
-			}
-		}
+	if delta == 0 {
+		// No need to do a db write for a no-change operation.
+		return c.value, true, nil
 	}
 
 	c.value += delta
-	// Create a lease
-	ttl := time.Until(next.Add(c.ttlOffset))
-	lease, err := c.etcdclient.Grant(ctx, int64(ttl.Seconds()))
+
+	_, err := c.etcdclient.KV.Put(ctx, c.key, strconv.Itoa(c.value))
+	return c.value, true, err
+}
+
+func (c *etcdcounter) Delete(ctx context.Context) error {
+	_, err := c.etcdclient.KV.Delete(ctx, c.key)
+	return err
+}
+
+func (c *etcdcounter) Refresh(ctx context.Context) (int, error) {
```

Comment:
Why is this called `Refresh` if its a etcd `Get` -Should it just be called `Get`? For ex, above you have a Delete func and you are performing an etcd `Delete`, which is intuitive, but `Refresh` is less explicit.

Reply:
This is to force a fetch from the DB. I can call it Fetch. Get might give the perception of a cache hit while this one skips the cache.

Reply:
Or Reload?

Reply:
ah ok, Refresh works then 👍🏻